### ISSUE TITLE
Android Studio

### ISSUE THREAD

**Issue by finneapps** [readium/SDKLauncher-Android#61]
Hi. Since all new Android project will be developed in Android Studio - it would be nice if Readium could work with gradle/Studio. Has anyone managed to make Readium work within Android Studio? Im getting this error after importing the project:

/Users/Finne/Documents/projects/SDKLauncher-Android-Gradle/ePub3Library/src/main/jni/container.cpp:22:27: fatal error: ePub3/archive.h: No such file or directory
 #include <ePub3/archive.h>
                           ^
compilation terminated.
make: *** [/Users/Finne/Documents/projects/SDKLauncher-Android-Gradle/ePub3Library/build/intermediates/ndk/release/obj/local/arm64-v8a/objs/ePub3Library//Users/Finne/Documents/projects/SDKLauncher-Android-Gradle/ePub3Library/src/main/jni/container.o] Error 1
Error:Execution failed for task ':ePub3Library:compileReleaseNdk'.
> com.android.ide.common.process.ProcessException: org.gradle.process.internal.ExecException: Process 'command '/Applications/android-ndk-r10d/ndk-build'' finished with non-zero exit value 2

**Comment by danielweck** [comment]
I have not been able to setup a fully-functioning Gradle project including both the ReadiumSDK (C++ and JNI wrapper) and the actual Android "launcher" application. However, I think @JCCR had some degree of success (debugging into JNI code remains a problem, but editing the code and invoking the `ndk-compile.sh` build script is possible). I see that you are not using Windows, but for those who do: the build process is even more involved (Cygwin highly recommended). Personally, I use OSX as my primary development platform, but I do spend a lot of time in Windows too (so I would love a cleaner / easier cross-platform Readium-Android workflow).

**Comment by finneapps** [comment]
Hi Daniel, thanks for your (quick) answer.  I hope there is a solution, because going back to Eclipse is not really an option anymore. I tried Readium a year ago and it worked great, but since that I have moved on to Android Studio. My company is planing to test Readium on iOS and Android

**Comment by rkwright** [comment]
@finneapps : We do intend to migrate to Android Studio, but my understanding is that support for C++ and the NDK under Android Studio is less than stellar.  Do you have other information?  What is your experience?

**Comment by finneapps** [comment]
I havent tried it, but it seems possible:
http://shaneenishry.com/blog/2014/08/17/ndk-with-android-studio/

**Comment by finneapps** [comment]
was the link of any use or is it something else you need Android Studio to be able to do?

**Comment by rkwright** [comment]
@finneapps Haven't had time to look yet. I will ask the relevant members of the team to look into it!

**Comment by danielweck** [comment]
@finneapps I remember coming across this link, probably when I was researching the subject, at about the same time as @JCCR
Converting the Eclipse project to AndroidStudio / Gradle is just one aspect (dare I say, the most trivial part). The hard part is to offer a build workflow that is easy to run across Windows, OSX, Linux, etc., and as a bonus it would be nice if we were able to step into C++ / JNI code when debugging the Android application. Regarding the build workflow, at the moment `ndk-compile.sh` is a shell script that requires copying the header files, etc. (not straight-forward on Windows, where the file path conventions and the NDK_PATH environment parameters are not abstracted), it would be better if the native makefile was somehow more transparently-integrated with the cross-platform Gradle build definitions, so that the epub3.lib (time-consuming) rebuild would be selectively triggered when C++ or JNI code changes.

**Comment by finneapps** [comment]
I have been trying to export the epub library to a jar in Eclipse and then use it in Android Studio. I havent been able to make it work and I dont know if it is possible. Im getting the unsatisfied link error

**Comment by danielweck** [comment]
When you unzip the jar, does it contain the native lib for the correct target platform? (arm, etc)

**Comment by finneapps** [comment]
It works now. So the current solution is to generate a jar from Eclipse and then use the jar in my Android Studio project

**Comment by finneapps** [comment]
Android Studio 1.3 now supports NDK:
https://sites.google.com/a/android.com/tools/recent/androidstudio13preview1available

**Comment by danielweck** [comment]
@JCCR  (Juan Corona, EvidentPoint) brought this to my attention last week. 

The Android Studio NDK plugin is based on JetBrains' Clion platform ( https://www.jetbrains.com/clion ), C/C++ code editing *and* **debugging** is supported, but I have not tried it yet myself (it would be great if we could place breakpoints in the ReadiumSDK as well as in launcher-app Java code).

Best to try the Canary channel directly: http://tools.android.com/download/studio/canary

**Comment by danielweck** [comment]
Related: https://github.com/readium/readium-sdk/issues/180

**Comment by JCCR** [comment]
Looks like the latest build in canary (1.3 beta) still doesn't have the C/C++ support:
http://tools.android.com/recent/androidstudio13betaavailable

They do claim it will be out soon though.

**Comment by balaji24** [comment]
@finneapps: Could you please explain a little further about how you got the project to work in Android Studio? Creating a JAR solution sounds great and would perfectly satisfy my use case. But if you could explain exactly what you did, that would be super helpful.

Thank you.

**Comment by JCCR** [comment]
@balaji24

I'll post how I got the project on Android Studio. As to making a JAR out of it, I haven't tried. I'd be interested in seeing how @finneapps did this.

**Comment by balaji24** [comment]
@JCCR  that would be super helpful to the community at large. Thank you so much for taking the time to do this.

I'm hoping this doesn't seem rude but do you know when the post would be complete?

**Comment by JCCR** [comment]
@balaji24 
Sorry I was going to post something sooner but I was having some issues reproducing my previous steps.

But finally this is how I imported the project without having to do a gradle migration:

(Assuming you are using OSX, maybe Linux will work too)
1. `mkdir ~/root-project-location && cd ~/root-project-location`
2. `git clone --recursive https://github.com/readium/SDKLauncher-Android.git`
3. Quick Start menu > Import project (Eclipse ADT, Gradle, etc.)
4. Select ~/root-project-location/SDKLauncher-Android (not the inner SDKLauncher-Android folder)
5. Create project from existing sources > Next > Project name, location > Next
6. Next you select the source file directories. I like to uncheck the two Readium_SDK_Test_Android entries to simplify things (I don't use that project)
7 . Review the libraries found, and there should be a 'libs' entry with two contents> Next > Review suggested modules (Two modules named Android and SDKLauncher-Android)
8. Select project SDK. Here I select Android API 21 Platform > Next > Folders are scanned..
9. It then shows me two AndroidManifest.xml files in a list > Finish
10. While the project is indexing in Android Studio:
11. Open a terminal again and cd to ~/root-project-location/SDKLauncher-Android/readium-sdk/
12. `sh MakeHeaders.sh Android`
13. `cd Platform/Android`
14. `./ndk-compile.sh build`
15. This will compile the readium-sdk, if it is successful you will see: `[armeabi-v7a] Install        : libepub3.so => libs/armeabi-v7a/libepub3.so`
16. Go back to Android Studio, create a Run/Debug Configuration for Android Application, select the SDKLauncher-Android module.
17. You should be good to go.

Let me know if you have any issues following these steps :)

**Comment by balaji24** [comment]
@JCCR thank you so much for the above steps. I had a few questions:
1) For step 7, apart from the 'libs' library, there is also the epub3-library found. I need to include that too, correct?
2) Once I finish step 15 and go back to Android Studio, I get few popups:
  - Unregistered VCS roots detected
           The following directories are roots of VCS repositories, but they are not registered in the Settings: 
           /Users/Development/Android/Redium/SDKLauncher-Android/readium-sdk
           /Users/Development/Android/Redium/SDKLauncher-Android/readium-shared-js
           /Users/Development/Android/Redium/SDKLauncher-Android
  So I clicked on 'Add Roots' and that fixes this popup.
  - But the other popup is related with project migration to gradle. How do I go about fixing this issue?

Thank you once again for helping out with this.

**Comment by balaji24** [comment]
@JCCR thank you so much for the above steps. I had one more question:
After step 15, when I open up Android Studio and create a Run/Debug Configuration for Android Application, under the 'Before Launch' tab, do I remove the 'Gradle-aware Make' task?

If I do remove it, do I need to include anything else in that tab? Because as of now, my application starts but is stuck on the 'Waiting For Debugger' state.

Thank you so much for helping out with this.

**Comment by JCCR** [comment]
@balaji24

You shouldn't have to remove it. My assumption is that internally the IDE has some compatibility mode that can be used to handle Eclipse projects, but from the outside you will still see references to Gradle.

I tried debugging on my end and it does get past the 'Waiting For Debugger' state for me. Maybe the issue you are seeing is unrelated to the fact that it's an Eclipse project?

Does running without debugging work?

**Comment by JCCR** [comment]
@danielweck 
They released the NDK support now :)
http://tools.android.com/recent/androidstudiowithandroidndkpreviewsupportavailable

**Comment by balaji24** [comment]
@JCCR you were correct! Thank you so much.
I am able to RUN the app but not debug it.

If I debug with the 'Gradle-aware Make' task, then I get the following errors:
Error:Android Dex: [SDKLauncher-Android] Unable to execute DX
Error:Android Dex: [SDKLauncher-Android] com.android.dex.DexException: Multiple dex files define Lorg/readium/sdk/android/CFI;
Error:Android Dex: [SDKLauncher-Android] at com.android.dx.merge.DexMerger.readSortableTypes(DexMerger.java:596)
Error:Android Dex: [SDKLauncher-Android] at com.android.dx.merge.DexMerger.getSortedTypes(DexMerger.java:554)
Error:Android Dex: [SDKLauncher-Android] at com.android.dx.merge.DexMerger.mergeClassDefs(DexMerger.java:535)
Error:Android Dex: [SDKLauncher-Android] at com.android.dx.merge.DexMerger.mergeDexes(DexMerger.java:171)
Error:Android Dex: [SDKLauncher-Android] at com.android.dx.merge.DexMerger.merge(DexMerger.java:189)
Error:Android Dex: [SDKLauncher-Android] at com.android.dx.command.dexer.Main.mergeLibraryDexBuffers(Main.java:502)
Error:Android Dex: [SDKLauncher-Android] at com.android.dx.command.dexer.Main.runMonoDex(Main.java:334)
Error:Android Dex: [SDKLauncher-Android] at com.android.dx.command.dexer.Main.run(Main.java:277)
Error:Android Dex: [SDKLauncher-Android] at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Error:Android Dex: [SDKLauncher-Android] at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
Error:Android Dex: [SDKLauncher-Android] at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Error:Android Dex: [SDKLauncher-Android] at java.lang.reflect.Method.invoke(Method.java:606)
Error:Android Dex: [SDKLauncher-Android] at org.jetbrains.android.compiler.tools.AndroidDxRunner.runDex(AndroidDxRunner.java:161)
Error:Android Dex: [SDKLauncher-Android] at org.jetbrains.android.compiler.tools.AndroidDxRunner.main(AndroidDxRunner.java:294)
Error:Android Dex: [SDKLauncher-Android] at com.intellij.rt.execution.CommandLineWrapper.main(CommandLineWrapper.java:130)

**Comment by JCCR** [comment]
@balaji24 
This error seems familiar to me. Let me google some stuff to see how I can help you out.

**Comment by JCCR** [comment]
@balaji24 
If you update your build tools through the SDK manager does it help?

**Comment by balaji24** [comment]
@JCCR thank you so much for looking this up.

I do not see any updates for my build tools available.

**Comment by JCCR** [comment]
@balaji24 
OK so, most of what I found about that error boiled down to editing the build.gradle (which we can't in our use case) or something to do with duplicated library references.
http://stackoverflow.com/questions/21735493/com-android-dex-dexexception-multiple-dex-files-define

**Comment by balaji24** [comment]
@JCCR I just restarted the entire process from the beginning and now, I do not get any of those errors! Its working on both DEBUG and RUN.

Thank you sooooo much for the help. Really appreciate it!

**Comment by balaji24** [comment]
@JCCR quick question: If I continue to use this solution, do you know how I would go about integrating this particular project with other gradle built projects? Also, I do not see any resource/layout files to work with. How do I go about getting access to all that?

Maybe first import into Eclipse, make all the changes required and follow the above steps to import into Android Studio? Do you think that will work?

**Comment by JCCR** [comment]
@balaji24 
You're welcome :)

To integrate with other projects we might have to include the project as a JAR dependency similarly to how fineapps mentioned.

As for the access to resource files, switch to "Project" view from "Android" view in the file tree:
![screen shot 2015-07-16 at 9 36 13 am](https://cloud.githubusercontent.com/assets/5132652/8729135/27550276-2b9e-11e5-892a-5410a22d245f.png)

**Comment by finneapps** [comment]
@JCCR What I did was to make everything run in Eclipse first. After that I exported the epub3-library as a jar. I included that jar in my Android Studio project, but it did not work (unsatisfied link error). Then I read this post:
http://wudev.digitaltorque.com/2014/03/building-readium-for-x86-android.html
and downloaded this jar:
http://cl.ly/141r030z3S33

I then included both jars and then it work for some reason *magic*. Build times was very slow so I hope you will find a better way :-1:

**Comment by balaji24** [comment]
@finneapps so all you did was import the SDKLauncher-Android project into Android Studio and include the two jars? is that all you needed to make it work?

**Comment by gmale** [comment]
So I got this to run in Android Studio, essentially by running `ndk-compile.sh` to generate the `libepub3.so` which I placed in the `src/main/jniLibs/armeabi-v7a` folder of the `readium-sdk/Platform/android` project. From there, I created an AAR library out of it and added it as a dependency of `SDKLauncher-Android` (this could easily be further automated with gradle, but I skipped that for now).

With some minor tweaks, this all worked just fine and I got the SDKLauncher app to run on my phone. The problem is, once I select an ePub book from the list, it crashes with a cryptic error:

```
org.readium.sdk.android.launcher D/libepub3 [./../../Platform/Android/jni/epub3.cpp:248]﹕ initializeReadiumSDK(): initializing Readium SDK...
org.readium.sdk.android.launcher D/libepub3 [./../../Platform/Android/jni/epub3.cpp:256]﹕ initializeReadiumSDK(): initialization of Readium SDK finished
org.readium.sdk.android.launcher D/libepub3 [./../../Platform/Android/jni/epub3.cpp:427]﹕ EPub3.openBook(): path received is '/storage/emulated/0/epubtest/alice3-1.epub'
org.readium.sdk.android.launcher D/libepub3 [./../../Platform/Android/jni/epub3.cpp:436]﹕ OpenContainer() EXCEPTION: ArchiveXmlReader: Nil ArchiveReader supplied
```

For some reason, [the ArchiveXmlReader is Nil which seems to come from this line of code](https://github.com/readium/readium-sdk/blob/master/ePub3/ePub/archive_xml.cpp#L31). Is there anyone familiar enough with the C++ project who can shed light on what might be causing this?

If I can get passed this little speed bump, I'd be happy to create a branch that works in Android Studio.

**Comment by efeyc** [comment]
@gmale I also experienced same issue. For me the reason for this error was that there was no storage permission in the manifest file. 

<uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />

I believe you can also use READ_EXTERNAL_STORAGE.
Because of this, it didn't find the file in SD card folder. I also realised that if the filename (or path) is wrong, it also gets the same error.

I hope your issue would be the same and fixed like this.

**Comment by danielweck** [comment]
@finneapps this feature branch works great: https://github.com/readium/SDKLauncher-Android/tree/feature/gradle-experimental
(and it's totally cross-platform now, no more Windows headaches!)

**Comment by rkwright** [comment]
Please use the develop branch.  If you run into other problems, please start a new issue.
